src/epplibwrapper/utility/pool.py

```diff
@@ -0,0 +1,133 @@
+
+    def __init__(self, client, login, options: dict):
+        # For storing shared credentials
+        self._client = client
+        self._login = login
+
+        # Keep track of each greenlet
+        self.greenlets: List[gevent.Greenlet] = []
+
+        # Define optional pool settings.
+        # Kept in a dict so that the parent class,
+        # client.py, can maintain seperation/expandability
+        self.size = 1
+        if "size" in options:
+            self.size = options["size"]
+
+        self.exc_classes = tuple((TransportError,))
+        if "exc_classes" in options:
+            self.exc_classes = options["exc_classes"]
+
+        self.keepalive = None
+        if "keepalive" in options:
+            self.keepalive = options["keepalive"]
+
+        # Determines the period in which new
+        # gevent threads are spun up
+        self.spawn_frequency = 0.1
+        if "spawn_frequency" in options:
+            self.spawn_frequency = options["spawn_frequency"]
+
+        self.conn: deque = deque()
+        self.lock = BoundedSemaphore(self.size)
+
+        self.populate_all_connections()
+
+    def _new_connection(self):
+        socket = self._create_socket(self._client, self._login)
+        try:
+            connection = socket.connect()
+            return connection
+        except Exception as err:
+            message = f"Failed to execute due to a registry error: {err}"
+            logger.error(message, exc_info=True)
+            # We want to raise a pool error rather than a LoginError here
+            # because if this occurs internally, we should handle this
+            # differently than we otherwise would for LoginError.
+            raise PoolError(code=PoolErrorCodes.NEW_CONNECTION_FAILED) from err
+
+    def _keepalive(self, c):
+        """Sends a command to the server to keep the connection alive."""
+        try:
+            # Sends a ping to EPPLib
+            c.send(Hello())
+        except Exception as err:
+            message = "Failed to keep the connection alive."
+            logger.error(message, exc_info=True)
+            raise PoolError(code=PoolErrorCodes.KEEP_ALIVE_FAILED) from err
+
+    def _create_socket(self, client, login) -> Socket:
+        """Creates and returns a socket instance"""
+        socket = Socket(client, login)
+        return socket
+
+    def get_connections(self):
+        """Returns the connection queue"""
+        return self.conn
+
+    def kill_all_connections(self):
+        """Kills all active connections in the pool."""
+        try:
+            if len(self.conn) > 0:
+                gevent.killall(self.greenlets)
+
+                self.greenlets.clear()
+                self.conn.clear()
+
+                # Clear the semaphore
+                self.lock = BoundedSemaphore(self.size)
+            else:
+                logger.info("No connections to kill.")
+        except Exception as err:
+            logger.error("Could not kill all connections.")
```

Comment:
**comment** you have this as an error in pool_error, why not use that here? It seems that error was made and never used